server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateTests.java

```diff
@@ -177,80 +149,277 @@ ClusterChangedEvent generateCloseEvent(boolean masterEligible) {
+            if (action instanceof GatewayMetaState.KeepPreviousGeneration) {
+                assertThat(action.getIndex(), equalTo(notChangedIndex.getIndex()));
+                GatewayMetaState.Transaction tx = mock(GatewayMetaState.Transaction.class);
+                assertThat(action.execute(tx), equalTo(3L));
+                verifyZeroInteractions(tx);
+            }
+            if (action instanceof GatewayMetaState.WriteNewIndexMetaData) {
+                assertThat(action.getIndex(), equalTo(newIndex.getIndex()));
+                GatewayMetaState.Transaction tx = mock(GatewayMetaState.Transaction.class);
+                when(tx.writeIndex("freshly created", newIndex)).thenReturn(0L);
+                assertThat(action.execute(tx), equalTo(0L));
+            }
+            if (action instanceof GatewayMetaState.WriteChangedIndexMetaData) {
+                assertThat(action.getIndex(), equalTo(newVersionChangedIndex.getIndex()));
+                GatewayMetaState.Transaction tx = mock(GatewayMetaState.Transaction.class);
+                when(tx.writeIndex(anyString(), eq(newVersionChangedIndex))).thenReturn(3L);
+                assertThat(action.execute(tx), equalTo(3L));
+                ArgumentCaptor<String> reason = ArgumentCaptor.forClass(String.class);
+                verify(tx).writeIndex(reason.capture(), eq(newVersionChangedIndex));
+                assertThat(reason.getValue(), containsString(Long.toString(versionChangedIndex.getVersion())));
+                assertThat(reason.getValue(), containsString(Long.toString(newVersionChangedIndex.getVersion())));
+            }
+        }
+    }
+
+    private static class MetaStateServiceWithFailures extends MetaStateService {
+        private final int invertedFailRate;
+        private boolean failRandomly;
+
+        private <T> MetaDataStateFormat<T> wrap(MetaDataStateFormat<T> format) {
+            return new MetaDataStateFormat<T>(format.getPrefix()) {
+                @Override
+                public void toXContent(XContentBuilder builder, T state) throws IOException {
+                   format.toXContent(builder, state);
+                }
+
+                @Override
+                public T fromXContent(XContentParser parser) throws IOException {
+                    return format.fromXContent(parser);
+                }
+
+                @Override
+                protected Directory newDirectory(Path dir) {
+                    MockDirectoryWrapper mock = newMockFSDirectory(dir);
+                    if (failRandomly) {
+                        MockDirectoryWrapper.Failure fail = new MockDirectoryWrapper.Failure() {
+                            @Override
+                            public void eval(MockDirectoryWrapper dir) throws IOException {
+                                int r = randomIntBetween(0, invertedFailRate);
+                                if (r == 0) {
+                                    throw new MockDirectoryWrapper.FakeIOException();
```

Comment:
should we randomly throw other exception types?

Reply:
Since we're mocking calls to Lucene directory, I'm not sure if throwing other exception types exception IOException is the right thing to do.